### PR TITLE
add donation endpoint 28

### PR DESCRIPTION
**Pull request by pablo-bergonzi**

### CODE REVIEW

File: src/DonateTo.ApplicationCore/Entities/DonationRequestCategory.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace DonateTo.ApplicationCore.Entities
+{
+    public class DonationRequestCategory
```
Comment (nicolasgranata) [inline src/DonateTo.ApplicationCore/Entities/DonationRequestCategory.cs:6]: Should it inherit from EntityBase?

File: src/DonateTo.ApplicationCore/Entities/DonationRequestItemCategory.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace DonateTo.ApplicationCore.Entities
+{
+    public class DonationRequestItemCategory
```
Comment (nicolasgranata) [inline src/DonateTo.ApplicationCore/Entities/DonationRequestItemCategory.cs:6]: Should it inherit from EntityBase?

File: src/DonateTo.ApplicationCore/Entities/Entity.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace DonateTo.ApplicationCore.Entities
+{
+    public abstract class Entity
```
Comment (nicolasgranata) [inline src/DonateTo.ApplicationCore/Entities/Entity.cs:5]: I think we already have a class called EntityBase, maybe the Entity class should be removed.

File: src/DonateTo.ApplicationCore/Interfaces/IRepository.cs [hunk 4]
```diff
@@ -16,21 +17,33 @@ public interface IRepository<TEntity> where TEntity : class
...
         /// <summary>
         /// Get a list of entities async.
         /// </summary>
         /// <param name="filter">Filter</param>
-        /// <returns>Task of IEnumerable of TEntity.</returns>
-        Task<IEnumerable<TEntity>> GetAsync(Expression<Func<TEntity, bool>> filter);
+        /// <returns>Task of IQueryable of TEntity.</returns>
+        Task<IQueryable<TEntity>> GetAsync(Expression<Func<TEntity, bool>> filter);
+
+                /// <summary>
```
Comment (nicolasgranata) [inline src/DonateTo.ApplicationCore/Interfaces/IRepository.cs:30]: Remove the space before ///<summary

File: src/DonateTo.Infrastructure/Data/Repositories/DonationRepository.cs [hunk 5]
```diff
@@ -8,5 +13,43 @@ public class DonationRepository : EntityFrameworkRepository<Donation, DonateToDb
...
+            return await Task.FromResult(this.GetHydratedDonations()).ConfigureAwait(false);
+        }
+        ///<inheritdoc cref="IRepository{Donation}"/>
+        public override Donation Get(long id)
+        {
+            return this.GetHydratedDonations().FirstOrDefault(d => d.Id.Equals(id));
+        }
+        public override async Task<Donation> GetAsync(long id)
+        {
+            return await this.GetHydratedDonations().FirstOrDefaultAsync(d => d.Id.Equals(id));
```
Comment (nicolasgranata) [inline src/DonateTo.Infrastructure/Data/Repositories/DonationRepository.cs:41]: Add configureAwait(false)

File: src/DonateTo.Infrastructure/Data/Repositories/DonationRequestRepository.cs [hunk 6]
```diff
@@ -8,5 +13,43 @@ public class DonationRequestRepository : EntityFrameworkRepository<DonationReque
...
+            return await Task.FromResult(this.GetHydratedDonationRequests()).ConfigureAwait(false);
+        }
+        ///<inheritdoc cref="IRepository{DonationRequest}"/>
+        public override DonationRequest Get(long id)
+        {
+            return this.GetHydratedDonationRequests().FirstOrDefault(d => d.Id.Equals(id));
+        }
+        public override async Task<DonationRequest> GetAsync(long id)
+        {
+            return await this.GetHydratedDonationRequests().FirstOrDefaultAsync(d => d.Id.Equals(id));
```
Comment (nicolasgranata) [inline src/DonateTo.Infrastructure/Data/Repositories/DonationRequestRepository.cs:41]: Add configureAwait(false)

File: src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        
+        /// <summary>
+        ///    Searches for a query string amongst multiple tables and multiples columns asynchronously.
+        /// </summary>
+        /// <param name="queryString"> String to search <param>
+        /// <param name="page"> Curent results page <param>
+        /// <param name="pageSize"> Size of results page <param>
+        /// <returns>Task of Paged DonationRequests of matching criteria.</returns>
+        public Task<PagedResult<DonationRequest>> SearchDonationRequestAsync(string queryString, int page, int pageSize) {
+           return SearchDonationRequestQuery(queryString).GetPagedAsync(page, pageSize );
```
Comment (nicolasgranata) [inline src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs:60]: should we add an await and configureAwait(false) here?

File: src/DonateTo.Services/Services/BaseService.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        ///<inheritdoc cref="IBaseService{TEntity}"/>
+        public virtual TEntity Get(long id) {
+            return this._entityRequestRepository.Get(id);
+        }
+
+         ///<inheritdoc cref="IBaseService{TEntity}"/>
+        public virtual async Task<TEntity> GetAsync(long id){
+            return await this._entityRequestRepository.GetAsync(id);
```
Comment (nicolasgranata) [inline src/DonateTo.Services/Services/BaseService.cs:37]: add configureAwait(false)

File: src/DonateTo.Services/Services/BaseService.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        ///<inheritdoc cref="IBaseService{TEntity}"/>
+        public virtual async Task<TEntity> UpdateAsync(TEntity entity, long id) {
+            if(entity.Id != id) {
+                throw new InvalidOperationException();
```
Comment (nicolasgranata) [inline src/DonateTo.Services/Services/BaseService.cs:43]: could we have a message in this exception?

File: src/DonateTo.Services/Services/DonationService.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace DonateTo.Services
+{
+    public class DonationService: BaseService<Donation>
+    {
+        public DonationService(IRepository<Donation> donationRepository):base(donationRepository)
+        {
+        }
+
+
```
Comment (nicolasgranata) [inline src/DonateTo.Services/Services/DonationService.cs:18]: Remove whitespaces

File: src/DonateTo.WebApi/V1/Controllers/BaseApiController.cs [hunk 11]
```diff
@@ -74,12 +83,13 @@ public virtual async Task<IActionResult> Post([FromBody] T value)
...
         /// <returns>Status 200 if the request has succeeded or
         ///  Status 500 if that have an error</returns>
-        [HttpPut("{id}")]
+        [HttpPut("{id}",Name = "[controller]_[action]")]
         [ProducesResponseType(StatusCodes.Status200OK)]
         [ProducesResponseType(StatusCodes.Status500InternalServerError)]
         public virtual async Task<IActionResult> Put(long id, [FromBody] T value)
         {
             var result = await _baseService.UpdateAsync(value, id).ConfigureAwait(false);
+            await _unitOfWork.SaveAsync();
```
Comment (nicolasgranata) [inline src/DonateTo.WebApi/V1/Controllers/BaseApiController.cs:92]: Add ConfigureAwait(false)

File: src/DonateTo.ApplicationCore/Entities/DonationRequestCategory.cs (hunk 1 above)
Comment (pablo-bergonzi) [reply src/DonateTo.ApplicationCore/Entities/DonationRequestCategory.cs:6]: I don´t think so, this isn't an actual Entity its only an artifact to go around a EF limitation, besides EntityBase has an ID while the FK's make up the composite key this uses. The only reason to inherit form there would be to know when and by who the relationship was created, though It seems a bit of an overkill.

File: src/DonateTo.ApplicationCore/Entities/DonationRequestItemCategory.cs (hunk 2 above)
Comment (pablo-bergonzi) [reply src/DonateTo.ApplicationCore/Entities/DonationRequestItemCategory.cs:6]: the same as the other rel entity

File: src/DonateTo.ApplicationCore/Entities/Entity.cs (hunk 3 above)
Comment (pablo-bergonzi) [reply src/DonateTo.ApplicationCore/Entities/Entity.cs:5]: you are right, this is a merge error, I'll delete it

File: src/DonateTo.Services/Services/BaseService.cs (hunk 9 above)
Comment (pablo-bergonzi) [reply src/DonateTo.Services/Services/BaseService.cs:43]: Sure

File: src/DonateTo.ApplicationCore/Entities/SampleModel.cs [hunk 12]
```diff
@@ -1,7 +0,0 @@
-﻿namespace DonateTo.ApplicationCore.Entities
-{
-    public class SampleModel : EntityBase
```
Comment (nicolasgranata) [inline src/DonateTo.ApplicationCore/Entities/SampleModel.cs:3]: Is this no longer used in the Sample Controller?

File: src/DonateTo.Infrastructure/Data/EntityFramework/EntityFrameworkRepository.cs [hunk 13]
```diff
@@ -16,7 +15,7 @@ namespace DonateTo.Infrastructure.Data.EntityFramework
     /// </summary>
     /// <typeparam name="TEntity"></typeparam>
     /// <typeparam name="TContext"></typeparam>
-    public abstract class EntityFrameworkRepository<TEntity, TContext> : IRepository<TEntity> where TEntity : class where TContext : DbContext
+    public abstract class EntityFrameworkRepository<TEntity, TContext> : IRepository<TEntity> where TEntity : class  where TContext : DbContext
```
Comment (nicolasgranata) [inline src/DonateTo.Infrastructure/Data/EntityFramework/EntityFrameworkRepository.cs:18]: I think this change is just an empty space :)

File: src/DonateTo.Services/Services/SearchService.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        ///<inheritdoc cref="ISearchService"/>
+        public PagedResult<DonationRequest> SearchDonationRequest(string queryString, int page, int pageSize) {
+            return this._searchRepository.SearchDonationRequest(queryString, page, pageSize);
+        }
+
+        ///<inheritdoc cref="ISearchService"/>
+        public Task<PagedResult<DonationRequest>> SearchDonationRequestAsync(string queryString, int page, int pageSize) {
+             return this._searchRepository.SearchDonationRequestAsync(queryString, page, pageSize);
```
Comment (nicolasgranata) [inline src/DonateTo.Services/Services/SearchService.cs:28]: Missing await and ConfigureAwait(false)

File: src/DonateTo.ApplicationCore/Interfaces/Services/IBaseService.cs [hunk 15]
```diff
@@ -61,13 +76,13 @@ public interface IBaseService<T> where T : class
...
         /// </summary>
-        /// <param name="id">Entity id.</param>
+        /// <param name="id">Id of Entity to delete.</param>
         void Delete(long id);
 
         /// <summary>
         /// Delete an entity async.
         /// </summary>
-        /// <param name="id">Entity id.</param>
+        /// <param name="entiidty">Id of Entity to delete.</param>
```
Comment (guillemurano) [inline src/DonateTo.ApplicationCore/Interfaces/Services/IBaseService.cs:85]: Name must match with parameter

File: src/DonateTo.ApplicationCore/Interfaces/Services/ISearchService.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+using DonateTo.ApplicationCore.Entities;
+using System.Threading.Tasks;
+
+namespace DonateTo.ApplicationCore.Interfaces.Services
+{
+    public interface ISearchService
+    {
+
+        /// <summary>
+        ///     Searches for a query string amongst multiple tables and multiples columns synchronously.
```
Comment (guillemurano) [inline src/DonateTo.ApplicationCore/Interfaces/Services/ISearchService.cs:11]: remove empty space leading summary description

File: src/DonateTo.Services/Services/SearchService.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace DonateTo.Services
+{
+    public class SearchService: ISearchService
+    {
+        private readonly IRepository<DonationRequest> _donationRepository;
+        private readonly ISearchRepository _searchRepository;
+        public SearchService(IRepository<DonationRequest> donationRepository, ISearchRepository searchRepository)
+        {
+            this._searchRepository = searchRepository;
```
Comment (guillemurano) [inline src/DonateTo.Services/Services/SearchService.cs:18]: 'this' is not needed, please remove

File: src/DonateTo.Services/Services/DonationRequestService.cs [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+using System.Collections.Generic;
+
+namespace DonateTo.Services
+{
+    public class DonationRequestService: BaseService<DonationRequest>
+    {
+        private readonly IRepository<DonationRequest> _donationRequestRepository;
+        public DonationRequestService(IRepository<DonationRequest> donationRequestRepository): base(donationRequestRepository)
+        {
+            this._donationRequestRepository = donationRequestRepository;
```
Comment (guillemurano) [inline src/DonateTo.Services/Services/DonationRequestService.cs:17]: 'this' is not needed, please remove

File: src/DonateTo.Services/Services/BaseService.cs [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+using System.Collections.Generic;
+
+namespace DonateTo.Services
+{
+    public abstract class BaseService<TEntity>: IBaseService<TEntity> where TEntity: EntityBase
+    {
+        private readonly IRepository<TEntity> _entityRequestRepository;
+        public BaseService(IRepository<TEntity> entityRequestRepository)
+        {
+            this._entityRequestRepository = entityRequestRepository;
```
Comment (guillemurano) [inline src/DonateTo.Services/Services/BaseService.cs:17]: 'this' is not needed, please remove

File: src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+using DonateTo.Infrastructure.Data.Extensions;
+using Microsoft.EntityFrameworkCore;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DonateTo.Infrastructure.Data.Repositories
+{
+    public class PostgresSearchRepository : ISearchRepository
+    {
+        private readonly DonateToDbContext DbContext;
```
Comment (guillemurano) [inline src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs:14]: Favor _dbContext as private readonly convention

File: src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+        public PostgresSearchRepository(DonateToDbContext dbContext)
+        {
+            this.DbContext = dbContext;
```
Comment (guillemurano) [inline src/DonateTo.Infrastructure/Data/Repositories/PostgresSearchRepository.cs:18]: 'this' is not needed, please remove

File: src/DonateTo.Infrastructure/Data/Repositories/DonationRequestRepository.cs [hunk 22]
```diff
@@ -8,5 +13,43 @@ public class DonationRequestRepository : EntityFrameworkRepository<DonationReque
...
+        private IQueryable<DonationRequest> GetHydratedDonationRequests() {
+            return DbContext.Set<DonationRequest>().Include( d => d.Address).Include( d => d.Status)
+                .Include( d => d.DonationRequestItems).Include( d => d.DonationRequestCategories)
+                .Include( d => d.Organization);
+        }
+
+        ///<inheritdoc cref="IRepository{DonationRequest}"/>
+        public override IQueryable<DonationRequest> Get()
+        {
+            return this.GetHydratedDonationRequests();
```
Comment (guillemurano) [inline src/DonateTo.Infrastructure/Data/Repositories/DonationRequestRepository.cs:26]: 'this' is not needed, please remove

File: src/DonateTo.Infrastructure/Data/Repositories/DonationRepository.cs [hunk 23]
```diff
@@ -8,5 +13,43 @@ public class DonationRepository : EntityFrameworkRepository<Donation, DonateToDb
...
+        private IQueryable<Donation> GetHydratedDonations() {
+            return DbContext.Set<Donation>().Include( d => d.Address).Include( d => d.Status).Include( d => d.DonationItems)
+                .Include( d => d.DonationRequest.DonationRequestItems).Include( d => d.DonationRequest.DonationRequestCategories)
+                .Include( d => d.DonationRequest.Organization).Include( d => d.DonationRequest.Status);
+        }
+
+        ///<inheritdoc cref="IRepository{Donation}"/>
+        public override IQueryable<Donation> Get()
+        {
+            return this.GetHydratedDonations();
```
Comment (guillemurano) [inline src/DonateTo.Infrastructure/Data/Repositories/DonationRepository.cs:26]: 'this' is not needed, please remove

**Pull request merged**